### ISSUE TITLE
Clicking on site notifications: "Unexpected error while performing an AJAX call. Maybe the connection to the server is lost."

### ISSUE THREAD

**Issue by iionly** [Elgg/Elgg#8413]
This error is thrown each time you click on a site notification (link).

**Comment by mrclay** [comment]
What version? Any change of interfering plugins?

**Comment by iionly** [comment]
* Happens on Elgg 1.10 and 1.11,
* does not happen on Elgg 1.0,
* I don't think 3rd party plugins are interferring (tested on Elgg 1.11.2 without any enabled),
* Tested on local site (installed in subdirectory) in case that matters - but if this would be the reasons I would expect it to work without error message showing up all the time,
* I think other members at the Community site mentioned this error showing up recently, too. I don't remember the thread where it was mentioned though at the moment. It might have occured not with site notifications necessarily.

**Comment by juho-jaakkola** [comment]
Are there errors in the error log?

**Comment by iionly** [comment]
Sorry. No errors in the log or console.

**Comment by mrclay** [comment]
Can you send us the the Ajax request and response that's generating this?

**Comment by iionly** [comment]
The ajax request is from elgg.site_notifications.auto_delete in mod/site_notifications/views/default/js/site_notifications.php

It seems the error message gets triggered by
```php
if (!elgg.isFunction(options.error)) {
		// add a generic error handler
		options.error = function(xhr, status, error) { 
			elgg.ajax.handleAjaxError(xhr, status, error);
		};
}
```
in js/lib/ajax.js. This code is not yet included in Elgg 1.9 (so no error) but only included since 1.10.

Interesting to note: the error occurs on Firefox but not on Chrome.

For testing I just create a blog. Another user that receives an on-site notification on blog creation then gets the error when clicking on the notification to visit the blog due to auto-deletion of the notification.

**Comment by Pasley70** [comment]
In our case this happens only in Firefox (38.0.5) for Windows (not Mac). We have no problems with other browsers. And we have some more Firefox-specific issues with AJAX, for example the ckeditor will only shown after reload the page (caching?).

**Comment by iionly** [comment]
Possibly yet another ajax issue on Elgg 1.11 (and 1.10?): https://community.elgg.org/discussion/view/2222105/ajax-error-in-elgg-1112.

**Comment by iionly** [comment]
Yet more: https://community.elgg.org/discussion/view/2235828/unexpected-error-while-performing-an-ajax-call-maybe-the-connection-to-the-server-is-lost

**Comment by iionly** [comment]
Now I have this problem also with the Elggchat plugin on Elgg 2.0 beta2 when closing a chat session. The error message does only show up on Elgg 2 and not on Elgg 1.10 / 1.11 (not tested on 1.12). I don't get any entries in the logs or browser console at all that would indicate where the problem is.

It might also be interesting to note that the chat session will only be closed after a full page reload.

This is becoming a very annoying problem!

**Comment by iionly** [comment]
It seems it is mandatory to provide both "success" and "error" options and to return success / error from the action to avoid the "Unexpected error while performing an AJAX call. Maybe the connection to the server is lost." error message.

Is it anywhere documented what Elgg exactly expects?

**Comment by iionly** [comment]
More testing...

The error option seems not to be the deciding factor for the error to show up. But it seems - at least for this very action (leave action in elggchat) it is necessary to have
```php
$response = array('success' => true);
echo json_encode($response);
```
added even if the response is not even dealt with in the success function.

I don't know if the error is only showing up for this action and in this plugin for this very reason. But it would be quite a bummer if it would...

As the same error seems also to occur on Elgg 1.10 onwards already at least in some occasions I wonder if this isn't a BC breaking issue even.

**Comment by iionly** [comment]
For the error when clicking on a site notification the reason is no error/success handling within the function elgg.site_notifications.auto_delete() in /mod/site_notifications/views/default/site_notifications.js and the delete action of the site_notification plugin respectively.

Any suggestions / solutions how the correct handling for the site notifications deletion should be implemented? This would fix the problem at least for site notifications.

But the same problem might occur in a lot of other cases where no custom success / error handling is implemented. In all these cases there would be a fallback to the generic error handler in /js/lib/ajax.js that was introduced in Elgg 1.10. Is this really as it is supposed to be?

**Comment by mrclay** [comment]
Fixed by #8994
